src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -247,30 +247,33 @@ protected override async Task<HttpResponseMessage> SendAsync (HttpRequestMessage
 			Volatile.Write (ref sentRequest, true);
 
 			var nsrequest = await CreateRequest (request).ConfigureAwait(false);
-			var dataTask = session.CreateDataTask (nsrequest);
-
-			var tcs = new TaskCompletionSource<HttpResponseMessage> ();
-
-			cancellationToken.Register (() => {
-				RemoveInflightData (dataTask);
-				dataTask?.Cancel ();
-				tcs.TrySetCanceled ();
-			});
+			NSUrlSessionDataTask dataTask = null;
 			
+			var tcs = new TaskCompletionSource<HttpResponseMessage> ();
+	
 			if (!cancellationToken.IsCancellationRequested) {
-				lock (inflightRequestsLock)
+				lock (inflightRequestsLock){
+					dataTask = session.CreateDataTask(nsrequest);
+					
 					inflightRequests.Add (dataTask, new InflightData {
 						RequestUrl = request.RequestUri.AbsoluteUri,
 						CompletionSource = tcs,
 						CancellationToken = cancellationToken,
 						Stream = new NSUrlSessionDataTaskStream (),
 						Request = request
 					});
-
-				if (dataTask.State == NSUrlSessionTaskState.Suspended)
+				}
+				
+				if (dataTask != null && dataTask.State == NSUrlSessionTaskState.Suspended)
```

Comment:
why the addition of `dataTask != null` ? have you run into that case ?

Reply:
it happens if cancellationToken.IsCancellationRequested which could theorically happen.

Reply:
no it can't since you moved that block (`if`) inside the condition (you mentioned)

Reply:
OMG you are right sorry for being this blind reading my own code !

Reply:
Is it ok for you @spouliot  ? How can I help you on this PR ? :)